bgpv1: graceful restart capability

harsimran-pabla opened the pull request:
#### Summary
Introducing BGP graceful restart configuration option per peer along with restart time parameter. This change includes setting this configuration via BGP CRDs and fetching the configuration via BGP CLI.

#### Behaviour change
Previously, BGP CP removes all peer configuration at shutdown. This part of the code is removed, because we do not want to send PEER_DECONFIGURED notification which results peer to tear down BGP session even if graceful restart was negotiated. 

#### RFC 4724 
This change supports only restarting speaker as described [here](https://github.com/osrg/gobgp/blob/master/docs/sources/graceful-restart.md#restarting-speaker). 

#### Example CRD
```
---
apiVersion: "cilium.io/v2alpha1"
kind: CiliumBGPPeeringPolicy
metadata:
  name: rack1
spec:
  nodeSelector:
    matchLabels:
      rack: rack1
  virtualRouters:
  - localASN: 65011
    exportPodCIDR: true
    neighbors:
    - peerAddress: "10.0.0.2/32"
      peerASN: 65011
      gracefulRestart:
        enabled: true
        restartTime: "20s"
```

#### CLI output 
```
cilium# cilium bgp peers -o json
[
  {
    "applied-hold-time-seconds": 9,
    "applied-keep-alive-time-seconds": 3,
    "configured-hold-time-seconds": 90,
    "configured-keep-alive-time-seconds": 30,
    "connect-retry-time-seconds": 120,
    "families": [
      {
        "accepted": 11,
        "advertised": 2,
        "afi": "ipv4",
        "received": 11,
        "safi": "unicast"
      },
      {
        "advertised": 1,
        "afi": "ipv6",
        "safi": "unicast"
      }
    ],
    "graceful-restart": {
      "enabled": true,
      "restart-time-seconds": 20
    },
    "local-asn": 65011,
    "peer-address": "10.0.0.2",
    "peer-asn": 65011,
    "session-state": "established",
    "uptime-nanoseconds": 46313329048
  }
]
```

#### PCAP

![Graceful Restart Open Message](https://github.com/cilium/cilium/assets/128612031/0c367985-0ae5-4810-ac5c-1fea46786e4a)

#### Raw PCAP


[graceful-restart.pcap.txt](https://github.com/cilium/cilium/files/11621908/graceful-restart.pcap.txt)

Fixes: https://github.com/cilium/cilium/issues/24256

```release-note
Add support for BGP graceful restart configuration via CiliumBGPPeeringPolicy CRD
```

harsimran-pabla committed:
bgpv1: introducing graceful restart capability

Introducing bgp graceful restart configuration option per peer along
with restart time parameter. This change includes setting this
configuration via BGP CRDs and fetching the configuration via BGP CLI.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

rastislavs commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumbgppeeringpolicies.yaml:
```diff
@@ -132,6 +132,18 @@ spec:
+                              take for the BGP session to be re-established with peer
+                              after a restart. After this period, peer will remove
+                              stale routes. This is described RFC 4724 section 4.2.
+                              \n Default is 120s if empty or zero."
```
I wonder why our default for `RestartTime` is `120s`:
 - RFC 4724, section 4 says "A suggested default for the Restart Time is a value less than or equal to the HOLDTIME carried in the OPEN."
- We default HoldTime to `90s`
- GoBGP defaults `RestartTime` to the [HoldTime value](https://github.com/osrg/gobgp/blob/f8c6d24e32d71a7ea0273a51b574fd63559e2943/internal/pkg/config/default.go#L224)

harsimran-pabla replied:
There was some discussion around this here : https://isovalent.slack.com/archives/C036D1MNKNG/p1682352752459779 

Conclusion was 120s default is in FRR and BIRD, so keeping things consistent with them.

rastislavs replied:
Got it, no major objections about that then.

rastislavs commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumbgppeeringpolicies.yaml:
```diff
@@ -132,6 +132,18 @@ spec:
                               whole second.
                             format: duration
                             type: string
+                          gracefulRestartEnable:
```
This is just my personal preference (feel free to ignore), but I would prefer using `gracefulRestartEnabled`, as CRD describes the desired state, not an action

rastislavs commented on api/v1/openapi.yaml:
```diff
@@ -3501,6 +3501,14 @@ definitions:
       uptime-nanoseconds:
         description: BGP peer connection uptime in nano seconds.
         type: integer
+      graceful-restart:
```
Just a personal preference, I would prefer `graceful-restart-enabled` more, also to be consistent with the CRD API.

rastislavs commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -92,6 +92,11 @@ func (g *GoBGPServer) GetPeerState(ctx context.Context) (types.GetPeerStateRespo
 
+		if peer.GracefulRestart != nil {
+			peerState.GracefulRestart = peer.GracefulRestart.Enabled
+			peerState.GracefulRestartRestartTimeSeconds = int64(peer.GracefulRestart.RestartTime)
```
Maybe it would also make sense to expose the [PeerRestartTime](https://github.com/osrg/gobgp/blob/f8c6d24e32d71a7ea0273a51b574fd63559e2943/api/gobgp.pb.go#L7560) in the state API?

harsimran-pabla replied:
hmm, gobgp does not set PeerRestartTime [here](https://github.com/osrg/gobgp/blob/master/internal/pkg/config/util.go#L550-L558) 

I will create upstream PR to fix it. In the meantime, we will just not display peer restart time - since it will be 0

harsimran-pabla replied:
https://github.com/osrg/gobgp/pull/2660

harsimran-pabla committed:
bgpv1: graceful restart unit tests

Added graceful restart unit tests in bgp-cp controller, reconciler and
state.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

rastislavs reviewed (approved):
LGTM now

harsimran-pabla committed:
bgpv1: include graceful restart in bgp-cp documentation

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

qmonnet commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumbgppeeringpolicies.yaml:
```diff
@@ -132,6 +132,24 @@ spec:
                             type: string
+                          gracefulRestart:
+                            description: GracefulRestart defined graceful restart
+                              parameters which are negotiated with this neighbor.
```
```suggestion
                            description: GracefulRestart defines graceful restart
                              parameters which are negotiated with this neighbor.
```

qmonnet commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -100,6 +117,11 @@ type CiliumBGPNeighbor struct {
 	// +kubebuilder:validation:Format=duration
 	KeepAliveTime metav1.Duration `json:"keepAliveTime,omitempty"`
+	// GracefulRestart defined graceful restart parameters which are negotiated
+	// with this neighbor.
```
```suggestion
	// GracefulRestart defines graceful restart parameters which are negotiated
	// with this neighbor.
```

qmonnet reviewed (approved):
I looked mostly at the doc changes, which are minimal and look good.

I spotted one typo in comments/description, please see below.

nathanjsweet reviewed (approved):
LGTM for K8s and API

YutaroHayakawa commented on api/v1/openapi.yaml:
```diff
@@ -3550,3 +3553,24 @@ definitions:
+      peer-restart-time-seconds:
+        description: |-
+          This is the restart time advertised by the peer.
+        type: integer
```
Are we willing to help peer's restart? Since we don't import the route from peers, maybe it's not so useful to do it. Then, probably it doesn't make sense to show this to users?

harsimran-pabla replied:
That's right, we do not really program any routes in FIB so we are not helper in this case. I will remove this part.

YutaroHayakawa commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -443,6 +438,11 @@ func (c *Controller) applyPolicyDefaults(policy *v2alpha1api.CiliumBGPPeeringPol
+
+			// apply graceful restart defaults
+			if n.GracefulRestart.Enabled && n.GracefulRestart.RestartTime.Duration == 0 {
+				n.GracefulRestart.RestartTime.Duration = types.DefaultGRRestartTime
```
RFC4724 Section 4 says 

```
A suggested default for the Restart Time is a value less than or equal to the HOLDTIME carried in the OPEN.
```

So, Restart Time should be dynamic, so we could delete the default constant, and it should be at least `n.GracefulRestart.RestartTime.Duration = n.HoldTime.Duration`?

harsimran-pabla replied:
I was debating what should be the default. [FRR](https://docs.frrouting.org/en/latest/bgp.html#clicmd-bgp-graceful-restart-restart-time-0-4095) uses 120s regardless of hold time. 

I think risk of making it equal to hold time is that cilium agent might not boot up within user defined hold time.

Default hold time is 90s, which is reasonable for restart time as well. 

But if user sets hold time too aggressive - like 3s, restart time of 3s is bit too short imo.

YutaroHayakawa replied:
Ah, indeed 💡 Then, no objection to the current value.

YutaroHayakawa commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -92,6 +92,13 @@ func (g *GoBGPServer) GetPeerState(ctx context.Context) (types.GetPeerStateRespo
+		if peer.GracefulRestart != nil {
+			peerState.GracefulRestart.Enabled = peer.GracefulRestart.Enabled
+			peerState.GracefulRestart.RestartTimeSeconds = int64(peer.GracefulRestart.RestartTime)
+			peerState.GracefulRestart.PeerRestartTimeSeconds = int64(peer.GracefulRestart.PeerRestartTime)
```
We can delete this if we agree that peer restart time is not needed.

YutaroHayakawa commented on pkg/bgpv1/types/bgp.go:
```diff
@@ -17,6 +17,9 @@ const (
 	DefaultBGPHoldTime = 90 * time.Second
+
+	// DefaultGRRestartTime defines default Restart Time for graceful restart (RFC 4724, section 4.2)
+	DefaultGRRestartTime = 120 * time.Second
```
We can delete this as I suggested in the above comment.

YutaroHayakawa commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -62,6 +62,23 @@ type CiliumBGPPeeringPolicySpec struct {
+	// After this period, peer will remove stale routes. This is
+	// described RFC 4724 section 4.2.
+	//
+	// Default is 120s if empty or zero.
```
We should mention that this only have second precision since we are rounding it up internally. Also, don't forget to update this default value when you agree with my suggestion and change the default.

harsimran-pabla replied:
I will circle back to this once we conclude on 120s GR default time.

YutaroHayakawa replied:
No need to change the default value, but still, the precision should be shown.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcile_test.go:
```diff
@@ -167,7 +177,8 @@ func TestNeighborReconciler(t *testing.T) {
 				{PeerASN: 64124, PeerAddress: "192.168.0.2/32"},
 			},
-			err: nil,
+			checks: checkTimers{},
```
We can just omit this since we're embedding the struct. Same for below.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcile_test.go:
```diff
@@ -1087,3 +1137,13 @@ func TestReconcileAfterServerReinit(t *testing.T) {
+	if addrNet.Is6() {
+		bits = 128
+	}
+	return netip.MustParsePrefix(fmt.Sprintf("%s/%d", addr, bits)).String()
```
You can use https://pkg.go.dev/net/netip#PrefixFrom here.

harsimran-pabla replied:
This is awesome!

YutaroHayakawa commented on pkg/bgpv1/manager/reconcile_test.go:
```diff
@@ -270,45 +322,43 @@ func TestNeighborReconciler(t *testing.T) {
+						Duration: time.Duration(peer.ConnectRetryTimeSeconds) * time.Second,
+					}
 				}
-			}
 
-			for _, p := range peers {
-				paddr := netip.MustParseAddr(p.PeerAddress)
-				var seen bool
-				for _, n := range tt.newNeighbors {
-					addr := netip.MustParsePrefix(n.PeerAddress)
-					if paddr == addr.Addr() {
-						seen = true
-						if n.ConnectRetryTime.Duration != 0 && int64(n.ConnectRetryTime.Seconds()) != p.ConnectRetryTimeSeconds {
-							t.Fatalf("ConnectRetryTime does not match: wanted: %d, got: %d", int64(n.ConnectRetryTime.Seconds()), p.ConnectRetryTimeSeconds)
-						}
+				if tt.checks.keepaliveTimer {
+					toCiliumPeer.KeepAliveTime = metav1.Duration{
+						Duration: time.Duration(peer.ConfiguredKeepAliveTimeSeconds) * time.Second,
```
It's a bit hard to follow from commit message what is the new test. I think this commit introduces 1) generic testing facilities for timers 2) graceful restart test using 1). So, please consider separating those two into different commits or describe them in detail in the commit message.

harsimran-pabla replied:
This refactor was to use require.ElementsMatch to compare two sets. I will split this into two commits

YutaroHayakawa commented on pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumbgppeeringpolicies.yaml:
```diff
@@ -133,7 +133,7 @@ spec:
                             type: string
                           gracefulRestart:
-                            description: GracefulRestart defined graceful restart
+                            description: GracefulRestart defines graceful restart
```
Please consolidate this fix to the original commit that introduced this description. Same for `bgpp_types.go`.

YutaroHayakawa commented on pkg/bgpv1/test/fixtures.go:
```diff
@@ -218,6 +218,8 @@ func setup(ctx context.Context, peerConfigs []gobgpConfig, fixConfig fixtureConf
 			peer.stopGoBGP()
 		}
 
+		f.bgp.BGPMgr.Stop()
```
Can't we hook this `Stop()` into the Hive's Stop lifecycle callback?

harsimran-pabla replied:
Including Stop in hive would mean that in prod code hive would call gobgp STOP while shutting down. This terminates all neighbors by sending PEER_DECONFIGURED, which results in peer not going into helper mode of graceful restart. 

So, calling this in test cleanup code - so we free up the BGP listen port.

YutaroHayakawa reviewed (changes requested):
Thanks for working for this! I made some comments for existing code. Also, I think we should elaborate more about which RFC/Draft we are referring in the commit message. At least I'm aware of the three GR-related standards that GoBGP implements; RFC4724 (The original GR), RFC8538 (NOTIFICATION support), draft-ietf-idr-long-lived-gr (LLGR). Also, it's hard to say we are compliant to the RFC without more evidence (e.g. PCAP of the exchanged BGP messages). Could you collect that?

harsimran-pabla committed:
bgpv1: refactor reconcile neighbor test for easier validation

Refactor bgpv1 reconcile neighbor test to match peers using testify
ElementsMatch. This reduces complexity of comparing different neighbor
parameters.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

YutaroHayakawa reviewed (approved):
Ok, that makes sense to me. Thanks for your clarification. Reordering the reconcilers addressed my last concern. Now looks good to me! Thanks for all of your effort to make this PR land including developing the testing framework!

YutaroHayakawa reviewed (approved):
Ok, that makes sense to me. Thanks for your clarification. Reordering the reconcilers addressed my last concern. Now looks good to me! Thanks for all of your effort to make this PR land including developing the testing framework!
